common/src/main/java/whocraft/tardis_refined/common/blockentity/device/AstralManipulatorBlockEntity.java

```diff
@@ -0,0 +1,261 @@
+        super(BlockEntityRegistry.ASTRAL_MANIPULATOR.get(), blockPos, blockState);
+    }
+
+    @Override
+    protected void saveAdditional(CompoundTag compoundTag) {
+        super.saveAdditional(compoundTag);
+
+        compoundTag.putBoolean(SHOULD_DISPLAY, this.shouldDisplay);
+
+        if (this.pointABlockPos != null) {
+            compoundTag.put(MAN_POINT_A, NbtUtils.writeBlockPos(this.pointABlockPos));
+        }
+
+        if (this.pointBBlockPos != null) {
+            compoundTag.put(MAN_POINT_B, NbtUtils.writeBlockPos(this.pointBBlockPos));
+
+        }
+    }
+
+    public BlockPos getPointABlockPos() {
+        return this.pointABlockPos;
+    }
+
+    public BlockPos getPointBBlockPos() {
+        return this.pointBBlockPos;
+    }
+
+    public void setShouldDisplay(boolean display) {
+        this.shouldDisplay = display;
+        this.pointABlockPos = this.getBlockPos();
+        this.pointBBlockPos = this.getBlockPos();
+        sendUpdates();
+    }
+
+    public boolean shouldDisplay() {
+
+        return shouldDisplay;
+    }
+
+    public void clearDisplay() {
+        this.shouldDisplay = false;
+        this.pointABlockPos = this.getBlockPos();
+        this.pointBBlockPos = this.getBlockPos();
+        sendUpdates();
+    }
+
+    @Override
+    public void load(CompoundTag tag) {
+
+        if (tag.contains(MAN_POINT_A)) {
+            this.pointABlockPos = NbtUtils.readBlockPos(tag.getCompound(MAN_POINT_A));
+        }
+        if (tag.contains(MAN_POINT_B)) {
+            this.pointBBlockPos = NbtUtils.readBlockPos(tag.getCompound(MAN_POINT_B));
+        }
+
+        if (tag.contains(SHOULD_DISPLAY)) {
+            this.shouldDisplay = tag.getBoolean(SHOULD_DISPLAY);
+        }
+
+        super.load(tag);
+    }
+
+
+    public void OnRightClick(ItemStack itemStack) {
+        if (itemStack.getItem() instanceof ScrewdriverItem screwdriverItem) {
+            if (!screwdriverItem.isScrewdriverMode(itemStack, ScrewdriverMode.DRAWING)) {
+                screwdriverItem.setScrewdriverMode(itemStack, ScrewdriverMode.DRAWING, getBlockPos(), null);
+                setShouldDisplay(true);
+                screwdriverItem.playScrewdriverSound((ServerLevel) getLevel(), getBlockPos(), SoundRegistry.SCREWDRIVER_CONNECT.get());
+
+            } else {
+
+                var points = screwdriverItem.getScrewdriverPoint(itemStack);
+                screwdriverItem.setScrewdriverMode(itemStack, ScrewdriverMode.DISABLED, getBlockPos(), null);
+
+                if (points.size() == 2) {
+                    var pointA = points.get(0);
+                    var pointB = points.get(1);
+
+                    attemptToBuild(pointA, pointB);
+                    screwdriverItem.clearBlockPosFromScrewdriver(itemStack);
+                    this.clearDisplay();
+                }
+
+                this.clearDisplay();
+            }
+        }
+    }
+
+    public boolean setBlockPos(BlockPos pos, boolean isPointA) {
```

Comment:
Perhaps use a different name? setPointBlockPos? The only reason I mention it is because setBlockPos sounds like a vanilla name for something else

Reply:
Renamed to setProjectionBlockPos